--- a/arch.h
+++ b/arch.h
@@ -87,6 +87,15 @@
 #    define AUDIT_ARCH_SPARC
 #    define MINIJAIL_ARCH_NAME "sparc"
 #  endif
+#elif defined(__loongarch64)
+/*
+ * AUDIT_ARCH_LOONGARCH64 may not be defined in older kernel headers.
+ */
+#  ifndef AUDIT_ARCH_LOONGARCH64
+#    define AUDIT_ARCH_LOONGARCH64 0xC0000102
+#  endif
+#  define MINIJAIL_ARCH_NR AUDIT_ARCH_LOONGARCH64
+#  define MINIJAIL_ARCH_NAME "loongarch64"
 #else
 #  error "AUDIT_ARCH value unavailable"
 #endif
--- a/util.c
+++ b/util.c
@@ -66,6 +66,8 @@
 const char *const log_syscalls[] = {"socket", "connect", "send"};
 #elif defined(__riscv)
 const char *const log_syscalls[] = {"socket", "connect", "sendto"};
+#elif defined(__loongarch64)
+const char *const log_syscalls[] = {};
 #else
 #error "Unsupported platform"
 #endif
